src/chains/ethereum/src/blockchain.ts

```diff
@@ -541,6 +561,267 @@ export default class Blockchain extends Emittery.Typed<BlockchainTypedEvents, Bl
+  public async traceTransaction(transactionHash:string, params:TransactionTraceOptions) {
+    const storageStack = {
+      currentDepth: -1,
+      stack: []
+    }
+
+    const returnVal = {
+      gas: 0,
+      returnValue: "",
+      structLogs: [] as Array<StructLog>
+    }
+
+    // #1 - get block via transaction object
+    const transaction = await this.transactions.get(transactionHash);
+
+    if (!transaction) {
+      throw new Error("Unknown transaction " + transactionHash);
+    }
+
+    const targetBlock = await this.blocks.get(transaction._blockNum);
+    const parentBlock = await this.blocks.getByHash(targetBlock.value.header.parentHash); 
+
+    // #2 - Set state root of original block
+    // 
+    // TODO: Forking needs the forked block number passed during this step: 
+    // https://github.com/trufflesuite/ganache-core/blob/develop/lib/blockchain_double.js#L917
+    let trie = new CheckpointTrie(this.#database.trie, parentBlock.value.header.stateRoot);
+
+    // Prepare the "next" block with necessary transactions
+    let newBlock = this.blocks.createBlock(parentBlock.value.header);
+
+    // make sure we use the same timestamp as the target block
+    newBlock.value.header.timestamp = targetBlock.value.header.timestamp;
+
+    for (const tx of targetBlock.value.transactions as Transaction[]) {
+      newBlock.value.transactions.push(tx)
+
+      // After including the target transaction, that's all we need to do.
+      if (Data.from(tx.hash()).toString() === transactionHash) {
+        break;
+      }
+    }
+
+    type StepEvent = {
+      gasLeft:BN;
+      memory:string;
+      stack:Array<BN>,
+      depth:number,
+      opcode: {
+        name: string
+      },
+      pc: number,
+      address:Buffer
+    }
+
+    // Note that this is async and 
+    let stepListener = (event:StepEvent, next:((error?:any, cb?:any) => void)) => {
+      // See these docs:
+      // https://github.com/ethereum/go-ethereum/wiki/Management-APIs
+
+      const gasLeft = event.gasLeft.toNumber();
+      const totalGasUsedAfterThisStep = Quantity.from(transaction.gasLimit).toNumber() - gasLeft;
+      const gasUsedPreviousStep = totalGasUsedAfterThisStep - returnVal.gas;
+      returnVal.gas += gasUsedPreviousStep;
+
+      let memory:string[] = null;
+      if (!params.disableMemory) {
+        let memoryAsString = Buffer.from(event.memory).toString("hex");
+        memory = memoryAsString.match(/.{1,64}/g) || [] as Array<string>;
+
+        if (memory.length > 0) {
+          const lastItem = memory[memory.length - 1];
+          if (lastItem.length < 64) {
+            memory[memory.length - 1] = lastItem + new Array(64 - lastItem.length + 1).join("0");
+          }
+        }
+      }
+
+      let stack:string[] = [];
+      if (!params.disableStack) {
+        for (const stackItem of event.stack) {
+          // Lots of conversions here: BN -> string -> Data -> string -> no-prefix string
+          // TODO: Make this more direct
```

Comment:
This string of conversions will likely get removed when I optimize this for memory (e.g., this is one of the spots where data is repeated throughout the response).

Reply:
This has been removed by later commits optimizing memory usage.